### PR TITLE
ADD python wrappers for pt/machine/recording info (WIP)

### PR DESCRIPTION
**Pull request by proloyd**
The libeep package does not expose a lot of header information, i.e., patient id, date of birth, machine info, channel types, start time etc etc. This pull request aims to incorporate extraction of such head info via libeep, so that those fields can be populated in the rawANT object. I am also happy to contribute to adding them in the mne.io.RawANT object (I could not find mne=1.9 where you are incorporating it.)

----
Great job @mscheltienne for making the python wrapper working! Our lab stumbled upon this a longtime ago, and we had to use their matlab importer as a work around. Let me know if you have any comments on the design.

### CODE REVIEW

**Comment by mscheltienne** [comment]
Thanks a lot for the addition, I will have a look to the code thius weekend or early next week!
Have a look at my comment in https://github.com/mne-tools/mne-python/issues/3609#issuecomment-2294792049

To keep the related PR smalls, I would like to merge this PR after https://github.com/mne-tools/mne-python/pull/12792 is merged; then we can open a second PR in MNE which modifies this initial reader to include the metadata you parsed here and depends on `antio` 0.3.0, released after this PR is merged.

Also, do you think it would be possible for you to record a couple of small files (couple of seconds) with your systems/amplifier setups/montage/eego version? It would be great to include tests files from different sites/conditions.

**Review by mscheltienne — APPROVED** [review]
LGTM, I pushed a couple of small commits. Let me know if you agree.
I could merge as is, but it would be great if we could include additional tests files which let's us test the channel status, type, the hospital, ... Would it be possible?

**Comment by proloyd** [comment]
Thanks for the changes,  looks good to me. I would like to ask you to wait for me to add additional test files. I should be able to get them by the end of this week. Thanks!

**Comment by mscheltienne** [comment]
Awesome!

For tests files, we decided to duplicate the dataset. It's not DRY but makes testing in `mne-python` and `antio` easier with little overhead. Tests files need to be added in:
- `antio` in `tests/data`
- https://github.com/mne-tools/mne-testing-data in `antio` (which duplicates `tests/data`)

If you want, I can handle the PR to `mne-testing-data` and `mne-python` to update `mne.datasets.testing` or can review/merge the PRs you will open.

**Comment by proloyd** [comment]
Sounds good! I can open PRs to `mne-testing-data` and `mne-python` to make things streamlined for you. 

I have couple of improvements to propose.
- Instead of C array -> list -> numpy array, we should directly convert the C array into numpy array for the sake of speed and memory efficiency. [I already implemented `get_samples_as_nparray()`, and tested with the current dataset] 
- Implement a `_read_segment_file()` that utilizes the above-mentioned function so that we don't have to preload data when reading the `.cnt` file. [TODO: should be easy.]

Lemme know what do you think!

**Comment by mscheltienne** [comment]
> Instead of C array -> list -> numpy array pathway, we should directly convert the C array into numpy array for the sake of speed and memory efficiency. [I already implemented get_samples_as_nparray(), and tested with the current dataset]
>  Implement a _read_segment_file() that utilizes the above-mentioned function so that we don't have to preload data when reading the .cnt file. [TODO: should be easy.]

Sure! I went the easy way with this first reader version, but both improvements would be very welcome. The second one will have to be done through a PR to `mne-python` directly. Thank you very much, I'll review in details today and will merge if you are happy with it!

**Comment by mscheltienne** [comment]
If we need numpy, we should build with the `2.0` version for compatibility with older and newer versions. 
I am not figuring out why I can't fix the build error, the header path seems correct..

File: pyproject.toml
Comment (mscheltienne) [inline pyproject.toml:97]: I don't think this should be needed as we already provide it in the `requires` of the `[build-system]`.

```suggestion
before-build = "pip install delvewheel"
```

**Comment by mscheltienne** [comment]
Looks like everything should come back green now. Have a look through my changes, let me know if you agree with them.

Let me know when you are happy for the PR and want it merge, LGTM!

**Comment by proloyd** [comment]
In C, typecasting always rounds down, i.e. works as floor(). Use of round() before the typecasting in exceldate to unixdate should take care of it.

**Comment by mscheltienne** [comment]
I really need to learn C 🙈 Thanks a lot for all the efforts you put in this early library!

File: src/libeep/python/pyeep.c
Comment (larsoner) [inline src/libeep/python/pyeep.c:230]: Is it 100% required for efficiency at this point to do this in C? I think having a C dependency on NumPy might make it so you can't use abi3 anymore (not 100% sure, though)

So if you have profiled and found that just using `list` here then converting to `ndarray` has a noticable performance impact then it might be worth doing this in C. But if you haven't profiled it might be worth it, keeping/doing this as a `list` and then casting to `np.array` in Python (rather than in C) could be better from a maintenance perspective

Comment (mscheltienne) [reply src/libeep/python/pyeep.c:230]: Arg! Good point, I think you are right and it won't be `abi3` anymore.
On windows, on a ~1h 128 channel file:
- list: 43.7 s ± 940 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
- numpy: 32.5 s ± 36.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

This is an extreme case with a long file, I don't think it's worth the maintenance trouble, it's the same order of magnitude..

Comment (larsoner) [reply src/libeep/python/pyeep.c:230]: You could return the `libeep_sample_data` buffer directly and `np.frombuffer` it in Pythotn, in theory that should be very similar to what the C code is actually doing and avoids the need for `list` at all.

Not 100% sure how to handle references and the like though but it might be worth trying to see if it is close to the C speed

Comment (proloyd) [reply src/libeep/python/pyeep.c:230]: >> having a C dependency on NumPy might make it so you can't use abi3 anymore.
This is probably not true. Also, the do's and don'ts are well documented here. https://numpy.org/doc/stable/dev/depending_on_numpy.html and a lot of package does rely on numpy-C APIs, so there will be constant support in this regard. @mscheltienne any comments?
 
>> So if you have profiled and found that just using list here then converting to ndarray has a noticable performance impact then it might be worth doing this in C. 
I did do profile before deciding on this. Using a `list`, then converting it to `ndarray` is ~30 times slower than creating the numpy array in C for a small (128x100) array. For longer data segments, it's gonna be even more.

Comment (proloyd) [reply src/libeep/python/pyeep.c:230]: > Arg! Good point, I think you are right and it won't be `abi3` anymore. On windows, on a ~1h 128 channel file:
> 
>     * list (`antio` 0.2.0): 43.7 s ± 940 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
> 
>     * numpy (`antio` 0.3.0.dev0 in this PR): 32.5 s ± 36.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
> 
> 
> This is an extreme case with a long file, I don't think it's worth the maintenance trouble, it's the same order of magnitude..

What exactly are you profiling? `cnt.get_samples()` vs `cnt.get_samples_as_nparray()` or `mne.io.read_raw_ant()` function calls? I find `cnt.get_samples_as_nparray()` to be ~25-30 times faster than `np.asanyaary(cnt.get_samples())`.

Comment (larsoner) [reply src/libeep/python/pyeep.c:230]: > I find cnt.get_samples_as_nparray() to be ~25-30 times faster than np.asanyaary(cnt.get_samples()).

It's going to depend probably on how much data you load at once. If you use a tiny array the Python call time overhead gets worse, so I'd expect your "small matrix" example to have a much worse overhead than what @mscheltienne used. So 25-30x slower for the tiny array vs < 2x slowdown for the "all data" test that @mscheltienne used doesn't sound unreasonable.

But hopefully this point is moot anyway if a `np.frombuffer` approach is used, I'd expect this to be very similar to the C code approach (at least for the "all data" loading... for tiny matrices there will still be some Python call overhead but I think that's okay).

Comment (proloyd) [reply src/libeep/python/pyeep.c]: Thanks for the idea @larsoner! Took a while to figure out how to return c-array as a python readable buffer, memoryview was the solution. The build is now numpy-free!

@mscheltienne I am not sure if we need the compat-test anymore. I kept it for you to take of it! 
It should be ready to merge after your review.

Comment (larsoner) [reply src/libeep/python/pyeep.c]: Awesome! 🚀

Comment (larsoner) [reply src/libeep/python/pyeep.c]: ... and @mscheltienne hopefully whatever benchmark you used before shows it's (nearly) as fast as the previous C approach for the whole-data-read case! 🤞

Comment (mscheltienne) [reply src/libeep/python/pyeep.c]: I was comparing `cnt. get_samples` and it's `numpy` version. The from buffer version was also my next idea, it's what is used in `mne-lsl`, it's awesome that you worked it out during my night 😂

I'll have a look today at the PR!

Comment (mscheltienne) [reply src/libeep/python/pyeep.c]: And performance similar to C: 

```
33.1 s ± 180 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

**Comment by mscheltienne** [comment]
Thanks @proloyd

File: src/libeep/python/pyeep.c
Comment (larsoner) [reply src/libeep/python/pyeep.c]: FYI in case anyone was curious, I was wrong -- you *can* use NumPy in ABI3 mode:

https://github.com/numpy/numpy/issues/27979
https://github.com/numpy/numpy/pull/27981

So if anything gets simpler/easier by reincluding NumPy, we can!

**Pull request merged**
